pkg/raft/term_cache_test.go

```diff
@@ -0,0 +1,396 @@
+				cache.truncateAndAppend(initialSlice)
+				b.StartTimer()
+				// This is the operation we're benchmarking
+				cache.truncateAndAppend(appendSlice)
+			}
+		})
+	}
+}
+
+func BenchmarkTermCacheScanAppendFewTermFlips(b *testing.B) {
+	// Create entries with few term flips
+	createEntriesWithFewFlips := func(count int, startIndex uint64, startTerm uint64) []pb.Entry {
+		entries := make([]pb.Entry, count)
+		currentTerm := startTerm
+		oneOverTermFlipFreq := 30
+
+		for i := 0; i < count; i++ {
+			// Only flip term every 30 entries
+			if i > 0 && i%oneOverTermFlipFreq == 0 {
+				currentTerm++
+			}
+			entries[i] = pb.Entry{
+				Index: startIndex + uint64(i),
+				Term:  currentTerm,
+			}
+		}
+		return entries
+	}
+
+	benchSizes := []int{10000, 100000, 1000000}
+	for _, size := range benchSizes {
+		b.Run(fmt.Sprintf("entries-%d", size), func(b *testing.B) {
+			initialEntries := createEntriesWithFewFlips(2, 1, 1)
+			initialSlice := LogSlice{prev: entryID{term: 5, index: 10}, entries: initialEntries}
+			entries := createEntriesWithFewFlips(size, 3, 3)
+			appendSlice := LogSlice{prev: entryID{term: 5 + 1, index: 10 + 1}, entries: entries}
+
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				b.StopTimer()
+				cache := NewTermCache(100, entryID{1, 1})
+				cache.truncateAndAppend(initialSlice)
+				b.StartTimer()
+
+				// This is the operation we're benchmarking
+				cache.truncateAndAppend(appendSlice)
+			}
+		})
+	}
+}
+
+func (tc *termCache) checkInvariants(t *testing.T) {
+	// check termCache is non-empty
+	require.NotEqual(t, 0, len(tc.cache), "termCache should never be empty")
+
+	// check termCache has increasing entries and terms
+	for i := 1; i < len(tc.cache); i++ {
+		if tc.cache[i].index <= tc.cache[i-1].index {
+			t.Errorf("termCache index is not strictly increasing: %d <= %d", tc.cache[i].index, tc.cache[i-1].index)
+		}
+		if tc.cache[i].term < tc.cache[i-1].term {
+			t.Errorf("termCache term is not monotonically increasing: %d < %d", tc.cache[i].term, tc.cache[i-1].term)
+		}
+	}
+}
+
+// generateRandomLogSlice creates a LogSlice with randomly generated entries for testing term cache behavior.
+// It simulates three scenarios based on random chance:
+//  1. Regular append: starts after endID
+//  2. Partial overwrite: starts between startID and endID
+//  3. Complete overwrite: starts before startID
+//
+// Parameters:
+//   - startID, endID: Define the range boundaries in the current term cache
+//   - averageLength: Controls entry count, generating between 1 and 2*averageLength entries
+//
+// The function ensures all generated entries maintain required term ordering invariants.
+func generateRandomLogSlice(startID, endID entryID, averageLength int) (lg LogSlice) {
+	if startID.index > endID.index || startID.term > endID.term {
+		panic("startID must be less than endID")
+	}
+	if startID.index == 0 || startID.term == 0 {
+		panic("startID must be higher than zero")
+	}
```

Comment:
In test helpers like this, it's common to pass in `t *testing.T`, and then use `require` to assert on things, instead of panic.